Merge integrate-pq into main

WillChilds-Klein opened the pull request:
### Issues:
Partially addresses CryptoAlg-1665

# Notes

There were a few issues in the merge that I had to manually resolve:

1. NID conflict on 973 for Dilithium3 and Kyber768. I resolved this by
   giving Dilithium3 an NID of 975 instead (Kyber1024 is 974).
   - After updating nid.h, I re-ran util/generate_build_files.py.
   - After updating files in crypto/obj, I reran the objects.go script.
2. Dilithium3's EVP_PKEY_ASN1_METHOD struct definition was slightly
   outdated and didn't have a priv_encode_v2 implementation. So, I gave
   it a trivial implementation that returns error, specifically EVP err
   EVP_R_OPERATION_NOT_SUPPORTED_FOR_THIS_KEYTYPE. We may want to come
   back and implement this method which "encodes |key| as a
   OneAsymmetricKey (RFC 5958)".

To simplify this PR's diff and facilitate reviewing, some portions of the merge
contents have been deleted in discrete commits. These deletions are temporary,
and will be reverted immediately post-merge in a two fast-follow PRs:

1. Reinstate Dilithium KATs and fuzzing corpus
2. Reinstate Dilithium documentation, and adjust the example code per
   @dkostic's guidance

Regarding the code remaining in this PR, much of it has been reviewed
previously. Here's a summary of the PRs for Dilithium-related code:

- PR #614 
- PR #674
- PR #724   
- PR #759 
- PR #805

# Testing
- CI checks
- local testing for the feature guard


Against test binary built with default build flags (i.e. `ENABLE_DILITHIUM=OFF`):

```
$ /build/crypto/crypto_test --gtest_filter='Dilithium3Test.*'
...
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from Dilithium3Test
[ RUN      ] Dilithium3Test.EvpDisabled
[       OK ] Dilithium3Test.EvpDisabled (0 ms)
[----------] 1 test from Dilithium3Test (0 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (0 ms total)
[  PASSED  ] 1 test.
```

Against test binary built with `-DENABLE_DILITHIUM=ON` set in cmake invocation:

```
$ /build/crypto/crypto_test --gtest_filter='Dilithium3Test.*'
...
[==========] Running 7 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 7 tests from Dilithium3Test
[ RUN      ] Dilithium3Test.KeyGeneration
[       OK ] Dilithium3Test.KeyGeneration (0 ms)
[ RUN      ] Dilithium3Test.KeyComparison
[       OK ] Dilithium3Test.KeyComparison (2 ms)
[ RUN      ] Dilithium3Test.NewKeyFromBytes
[       OK ] Dilithium3Test.NewKeyFromBytes (0 ms)
[ RUN      ] Dilithium3Test.KeySize
[       OK ] Dilithium3Test.KeySize (1 ms)
[ RUN      ] Dilithium3Test.Encoding
[       OK ] Dilithium3Test.Encoding (1 ms)
[ RUN      ] Dilithium3Test.Decoding
[       OK ] Dilithium3Test.Decoding (0 ms)
[ RUN      ] Dilithium3Test.SIGOperations
[       OK ] Dilithium3Test.SIGOperations (5 ms)
[----------] 7 tests from Dilithium3Test (9 ms total)

[----------] Global test environment tear-down
[==========] 7 tests from 1 test suite ran. (9 ms total)
[  PASSED  ] 7 tests.
```

---
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and 
the ISC license.

jakemas commented:
Thanks Will, I'll take a look.

On:
> NID conflict on 973 for Dilithium3 and Kyber768. I resolved this by
> giving Dilithium3 an NID of 975 instead (Kyber1024 is 974).
> After updating nid.h, I re-ran util/generate_build_files.py.
> After updating files in crypto/obj, I reran the objects.go script.

Perfect, yes that was an expected conflict that we planned to address during merge to main.

jakemas commented on crypto/dilithium/p_dilithium3.c:
```diff
@@ -0,0 +1,114 @@
+  }
+
+  if (DILITHIUM3_sign(sig, siglen, tbs, tbslen, key->priv) != 0) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_INVALID_SIGNATURE);
```
`EVP_R_INVALID_SIGNATURE` may be the wrong error to use here, as the signature is not strictly invalid (it likely just doesn't exist).

WillChilds-Klein replied:
`ERR_R_INTERNAL_ERROR` is used elsewhere in the file; I've gone with that for now.

jakemas commented on crypto/dilithium/pqcrystals-dilithium_dilithium_common/aes256ctr.h:
AES modes for Dilithium are unused, as only the SHA3 based Dilithium version is supported. This file is therefore unused and should be removed (it is an artefact of testing).

WillChilds-Klein replied:
done. i ripped out a bunch of other unused AES-related code as well.

WillChilds-Klein commented on crypto/dilithium/pqcrystals-dilithium_dilithium_common/aes256ctr.c:
done. i ripped out a bunch of other unused AES-related code as well.

jakemas reviewed (commented):
looks good to me!

dkostic commented:
Can we please rename the `pqcrystals-dilithium_dilithium_common` dir to `pqcrystals_dilithium_ref_common`.

dkostic commented on crypto/dilithium/README.md:
```diff
@@ -0,0 +1,71 @@
+* `rng.{h|c}` are deleted because we are using the randomness generation functions provided by AWS-LC.
+* `symmetric-aes.c` are removed because we are using only the SHA3 based Dilithium.
+* `sign.c`: calls to `randombytes` function is replaced with calls to `pq_custom_randombytes` and the appropriate header file is included (`crypto/rand_extra/pq_custom_randombytes.h`).
+* `sign.c` was modified from `*mlen = -1;` to `*mlen = 0;` within the `crypto_sign_open` function. This was modified as `*mlen = -1;` is attempting to assign `-1` to an unsigned integer, giving the warning: ` warning C4245: '=': conversion from 'int' to 'std::size_t'`. This warning would cause Windows x86 builds to fail. See [pq-crystals/dilithium#65](https://github.com/pq-crystals/dilithium/issu ...
```
not directly relevant to the PR, but @jakemas can you please ping the Dilithium maintainers about the opened issue?

jakemas replied:
Ok

dkostic commented on crypto/dilithium/README.md:
```diff
@@ -0,0 +1,71 @@
+EVP_PKEY_keygen(ctx, &dilithium_pkey)
+
+// Sign
+bssl::ScopedEVP_MD_CTX md_ctx;
```
- Nit: the example is mixing C and C++ style of using aws-lc. It'd be nice to settle on one (I'd prefer C).
- It would be good to add a full example of how to use the API (similarly to https://github.com/aws/aws-lc/tree/main/crypto/kem#full-example).

WillChilds-Klein replied:
I'm going to post a separate PR that will be solely concerned with dilithium documentation. we can continue this discussion there.

dkostic commented on crypto/dilithium/p_dilithium_test.cc:
```diff
@@ -0,0 +1,622 @@
+}
+
+TEST(Dilithium3Test, Decoding) {
+  static const uint8_t kPublicKey[] = {
```
can we move the constants outside of the test function please (for readability)?

WillChilds-Klein replied:
done.

dkostic commented on crypto/evp_extra/pq_sig_kat_tests_dilithium3.cc:
can we move this file and the KAT to the dilithium folder?

dkostic commented:
GIthub decided to stop showing me the diff starting from the fuzz folder to the end... so I'll try again tomorrow.

torben-hansen commented:
Can you split this into smaller pieces?

WillChilds-Klein commented:
> GIthub decided to stop showing me the diff starting from the fuzz folder to the end... so I'll try again tomorrow.

> Can you split this into smaller pieces?

the full diff is failing to render for me too. i'll try to delete portions of the merge piecemeal and then post additional PR's reverting those deletions.

WillChilds-Klein commented:
Ok, I've addressed feedback and have deleted a few things to make this PR's diff more manageable. Those deletions are temporary, and will be reverted immediately after this has merged. The fast-follow revert PRs will be:

1. Reinstate Dilithium KATs and fuzzing corpus
2. Reinstate Dilithium documentation, and adjust the example code per @dkostic's guidance

torben-hansen commented on crypto/dilithium/p_dilithium3_asn1.c:
```diff
@@ -0,0 +1,216 @@
+}
+
+static int dilithium3_priv_encode_v2(CBB *out, const EVP_PKEY *pkey) {
+OPENSSL_PUT_ERROR(EVP, EVP_R_OPERATION_NOT_SUPPORTED_FOR_THIS_KEYTYPE);
```
indentation

WillChilds-Klein replied:
will fix. we should consider onboarding a linter to AWS-LC's CI. s2n-tls uses clang-format, it's nice.

torben-hansen commented on crypto/dilithium/p_dilithium3_asn1.c:
```diff
@@ -0,0 +1,216 @@
+    dilithium3_pub_cmp,
+    dilithium3_priv_decode,
+    dilithium3_priv_encode,
+    dilithium3_priv_encode_v2,
```
It's caught in `EVP_marshal_private_key_v2()`. Leave it `NULL`.

WillChilds-Klein replied:
ah, nice. removed.

torben-hansen commented on crypto/dilithium/pqcrystals_dilithium_ref_common/config.h:
```diff
@@ -0,0 +1,37 @@
+
+#include <openssl/base.h>
+
+//#define DILITHIUM_MODE 2
```
?

WillChilds-Klein replied:
must be vestigial from jake's changes. i'll remove.

torben-hansen commented on crypto/dilithium/sig_dilithium.h:
```diff
@@ -0,0 +1,45 @@
+
+// DILITHIUM3_keypair generates a Dilithium3 keypair and assigns a public key to
+// |public_key| and a private key to |secret_key|. The function is a wrapper for
+// the keygen function from the Dilithium source. It returns 0 upon success.
```
returns what on error?

WillChilds-Klein replied:
_it doesn't [look like there's an error path in keygen](https://github.com/aws/aws-lc/blob/integrate-pq/crypto/dilithium/pqcrystals-dilithium_dilithium_common/sign.c#L11-L65)._

torben-hansen commented on crypto/dilithium/sig_dilithium.h:
```diff
@@ -0,0 +1,45 @@
+// output signature, |sig_len| is a pointer to output length of signature,
+// |message| is a pointer to message to be signed, |message_len| is the length
+// of the message and |secret_key| is a pointer to bit-packed secret key.
+// It returns 0 upon success.
```
returns what on error?

WillChilds-Klein replied:
it doesn't [look like there's an error path in signing](https://github.com/aws/aws-lc/blob/integrate-pq/crypto/dilithium/pqcrystals-dilithium_dilithium_common/sign.c#L67-178).

torben-hansen commented on crypto/dilithium/sig_dilithium.h:
```diff
@@ -0,0 +1,45 @@
+// to input signature, |sig_len| is the length of the signature, |message| is
+// a pointer to message, |message_len| is the length of the message and
+// |public_key| is a pointer to bit-packed public key. Returns 0 if signature
+// could be verified correctly and -1 otherwise.
```
```suggestion
// could be verified successfully and -1 otherwise.
```

WillChilds-Klein replied:
will do.

torben-hansen commented on crypto/dilithium/sig_dilithium3.c:
```diff
@@ -0,0 +1,40 @@
+/*
```
Use 
```
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
```

WillChilds-Klein replied:
will do.

torben-hansen commented on include/openssl/base.h:
```diff
@@ -438,6 +438,7 @@ typedef struct hmac_ctx_st HMAC_CTX;
 typedef struct md4_state_st MD4_CTX;
 typedef struct md5_state_st MD5_CTX;
 typedef struct ossl_init_settings_st OPENSSL_INIT_SETTINGS;
+typedef struct ossl_param_st OSSL_PARAM;
```
where does this come from?? This shouldn't be used in AWS-LC.

torben-hansen commented on crypto/x509/x509_test.cc:
```diff
@@ -2302,6 +2661,78 @@ TEST(X509Test, Ed25519Sign) {
 }
 
+TEST(X509Test, Dilithium3SignVerifyCert) {
+  //generate the dilithium key
```
Nit: I know it has been reviewed. But IMO, these comments are mostly redundant since it is obvious what is going on.

torben-hansen replied:
Now that some other stuff is changing anyway, could fix the indentation.

WillChilds-Klein replied:
sure, will fix indentation and remove particularly obvious comments.

WillChilds-Klein replied:
comments consolidate.

what's the issue with indentation? 2 spaces seems consistent with the rest of the file.

WillChilds-Klein commented on include/openssl/nid.h:
```diff
@@ -4299,6 +4299,10 @@ extern "C" {
 #define NID_KYBER1024_R3 974
 
+#define SN_DILITHIUM3 "DILITHIUM3"
+#define NID_DILITHIUM3 975
```
that's correct.

WillChilds-Klein commented on crypto/dilithium/p_dilithium3_asn1.c:
```diff
@@ -0,0 +1,216 @@
+}
+
+static int dilithium3_pub_decode(EVP_PKEY *out, CBS *params, CBS *key) {
+  // See https://datatracker.ietf.org/doc/draft-ietf-lamps-dilithium-certificates/ section 4.
```
this is a fair point. as suggested by @dkostic offline, perhaps we should consider adding a `_R3` suffix to the current dilithium stuff as we did for Kyber512 in our new KEM API.

@jakemas -- @torben-hansen hansen also had questions around the current state of standardization of dilithium in x509. is the current implementation/standard likely to change at all? if so, do we need to guard it with some sort of "experimental" qualifier or feature flag until the standard solidifies?

WillChilds-Klein replied:
added `_R3` suffix to the NID/OID as well as an opt-in build time feature guard.

WillChilds-Klein commented:
Ok, here are the requested changes:

- change NID name to `DILITHIUM3` to `DILITHIUM3_R3`
- a build-time feature guard for Dilithium3 (covers source, test, and benchmark)
    + see updated **Testing** section in PR overview for how the build flag affects Dilithium testing
- add Dilithium3 test CI run to the POSIX tests to ensure we always have full coverage with and without Dilithium enabled

Here's a log of commits I've made since responding to @torben-hansen's last round of feedback (the API version bump is unrelated to this PR and has already been merged):

```
$ git log --author childw d88c2ee..HEAD
commit 9f662afeec19cd7cdf8a9e65d013451177bd345d (HEAD -> cleaup-pq)
Merge: df834beba c48902790
Author: Will Childs-Klein <childw@amazon.com>
Date:   Mon May 8 19:06:29 2023 +0000

    Merge branch 'main' into cleaup-pq

commit 314c50d0054b52a71bab0227154809ea6ce5c206
Author: Will Childs-Klein <childw@amazon.com>
Date:   Mon May 8 17:57:56 2023 +0000

    Add disablement test when ENABLE_DILITHIUM=OFF

commit 7d08364fe91c11b491ee48c59d4a3cfe8ce6edd6
Author: Will Childs-Klein <childw@amazon.com>
Date:   Mon May 8 17:15:10 2023 +0000

    Rename DILITHIUM3 OID/NID to DILITHIUM3_R3

    This change qualifies the public DILITHIUM3 identifiers with a "round 3"
    (`_R3`) suffix. Internal function, variable, etc. are left as-is because
    they are not exposed to callers and can be updated if/when a materially
    different round 4 comes down the pipeline.

    Ideally, we'd be able to exclude this NID/OID entirely in non-dilithium
    builds, but this is not currently possible with the current
    `crypto/obj/` code. A next-best option would have been to put the
    Dilithium3 NID into an "experimental" NID namespace _well_ above the
    number of anticipated algorithms over the library's lifetime. I
    prototyped this out with an NID of 10,001 but as it turns out the
    tooling will expand NUM_NID and the `kObjects[]` in `obj_dat.h` and
    backfill all ~9000 ivervening array values with NULL/0-valued entries.
    This space tradeoff seemed unacceptable, so this implementation presents
    the `DILITHIUM3_R3` NID in all builds.

    The next commit will add a test to assert the behavior of trying to
    specify `DILITHIUM3_R3` in non-`DILITHIUM_ENABLED=ON` builds.

commit 51494fa7226e2080ebf2097960c8b64ca0a532bc
Merge: f3aa0ed9e 14daff9d3
Author: Will Childs-Klein <childw@amazon.com>
Date:   Fri May 5 17:05:27 2023 +0000

    Merge branch 'main' into cleaup-pq

commit f3aa0ed9e7b87d36869f5e9e294dc2f495776d25
Author: Will Childs-Klein <childw@amazon.com>
Date:   Fri May 5 17:04:10 2023 +0000

    Add Dilithium3 case to POSIX tests, enable dilithium in benchmarks

commit 8eeb737d15123fde4b6d213a82edf79690cae891
Author: Will Childs-Klein <childw@amazon.com>
Date:   Fri May 5 16:44:06 2023 +0000

    Fix speed.cc to work with new dilithium guard

commit ce5eda46933e9eae06425c721abe817c43f50a54
Author: Will Childs-Klein <childw@amazon.com>
Date:   Thu May 4 21:32:52 2023 +0000

    Add opt-in build feature guard for Dilithium

    Still need to:

    - change `DILITHIUM3` to `DILITHIUM3_R3` everywhere
    - move `crypto/dilithium/` to `crypto/dilithium3/r3`?
    - update CI to set this feature guard
    - perhaps move NID up to higher numerical value in an "experimental"
      namespace where it won't conflict with newer algos/NIDs?
    - #ifdef out the dilithium NID definition if possible

commit 5b88f4a09867612dc1aa040dcf2f52ee0a6fd2c8
Merge: b907f6b9c d723334d7
Author: Will Childs-Klein <childw@amazon.com>
Date:   Thu May 4 20:50:29 2023 +0000

    Merge branch 'main' into cleaup-pq

commit b907f6b9cd648085eb9b216fe4584f295de09e3a
Merge: 5d320e117 33741f13c
Author: Will Childs-Klein <childw@amazon.com>
Date:   Thu May 4 20:50:14 2023 +0000

    Merge branch 'main' into cleaup-pq

commit 28c85a29a45a69f37f8becb7cefe5faed96eb422
Author: Will Childs-Klein <childw@amazon.com>
Date:   Wed Apr 5 16:30:54 2023 -0400

    Bump AWS-LC API version (#900)

commit 5d320e117d74b547c7f944f9e159e78ad4fe545f
Merge: 0d0b0b937 a962ee3f3
Author: Will Childs-Klein <childw@amazon.com>
Date:   Tue Apr 4 19:13:05 2023 +0000

    Merge branch 'main' into cleaup-pq
```

As [mentioned previously](https://github.com/aws/aws-lc/pull/897#issuecomment-1492341382), once this PR has been merged there will be fast-follow PRs to:

1. Reinstate Dilithium KATs and fuzzing corpus
2. Reinstate Dilithium documentation, and adjust the example code per @dkostic's guidance

torben-hansen commented on tests/ci/run_posix_tests.sh:
```diff
@@ -13,6 +13,9 @@ build_and_test
 build_and_test -DCMAKE_BUILD_TYPE=Release
 
+echo "Testing AWS-LC with Dilithium3 enabled."
+build_and_test -DENABLE_DILITHIUM=ON
```
Is there another way? This runs the entire test suite again, duplicating everything except the dilithium tests.

WillChilds-Klein replied:
That's a fair question, and I agree that adding an additional test suite run is not ideal. Given that we're setting `ENABLE_DILITHIUM=OFF` by default, the testing options I'd identified were:

1. Run all POSIX tests with default `ENABLE_DILITHIUM=OFF`, set `ENABLE_DILITHIUM=ON` as one of the "uncommon build options" in the build-only checks below
2. Run all POSIX tests with non-default `ENABLE_DILITHIUM=ON`,  set `ENABLE_DILITHIUM=OFF` as one of the "uncommon build options" in the build-only checks below
3. Add an independent test run with `ENABLE_DILITHIUM=ON` (i.e. what this PR implements)

Option 1. has the distinct disadvantage of reducing Dilithium coverage to 0, as we'd only be doing compilation checks with the option set. Option 2. would retain Dilithium coverage, but leave the _default_ `ENABLE_DILITHIUM=OFF` build/execution path uncovered by testing, which seems dangerous. That left option 3., where we retain coverage over default and non-default configurations at the cost of (temporarily) increased test runtime.

Option 3. seemed like the safest of the available options.

WillChilds-Klein commented on crypto/dilithium/dilithium3_ref.c:
yep, will do.

torben-hansen reviewed (approved):
Prefer to squash this, otherwise the commit history would be mighty polluted.

For the squash commit message. Write something that makes sense. A high-level summary of changes maybe?
